File: pe018/pe018.py
#!/usr/bin/python

"""
Project Euler, Problems 18 & 67:

Find the maximum total from top to bottom in a text file containing a triangle
with:
 1) 15 (input18.txt) rows,
 2) 100 (input67.txt) rows.

https://projecteuler.net/problem=18
https://projecteuler.net/problem=67
"""

"""
Implementation note:
Denote x[k,i] as the i-th element in the k-th row, and total[k,i] as
the partial total from top to the i-th element in the k-th row.
The total[] can be evaluated as follows:
 * total[k, 0] = x[k, 0] + total[k-1, 0]
 * total[k, k] = x[k, k] + total[k-1, k-1]
 * total[k, i] = x[k, i] + max(total[k-1, i-1], total[k-1, i])

Note that to find total[k, *] we only need total[k-1, *], so there
is no need to keep all previous rows. Thus, memory footprint is
proportional only to the maximal length of a row.
"""

import sys

def main(filename):
    totals = []
    k = 0
    try:
        with open(filename, "r") as fi:
            for l in fi:
                k += 1
                if k == 1:
                    totals.append(0)
                else:
                    totals.append(totals[k - 2])
                    for i in range(k - 2, 0, -1):
                        totals[i] = max(totals[i], totals[i - 1])

                values = [ int(x) for x in l.split(' ') ]
                for i in range(0, k):
                    totals[i] += values[i]
            fi.close()
        print(max(totals))
    except:
        sys.stderr.write("Error: failed to read data!\n")
        sys.exit(1)

if __name__ == "__main__":
    main("input18.txt")
    main("input67.txt")

File: pe018/pe018.go
/*
Project Euler, Problems 18 & 67:

Find the maximum total from top to bottom in a text file containing a triangle
with:
 1) 15 (input18.txt) rows,
 2) 100 (input67.txt) rows.

https://projecteuler.net/problem=18
https://projecteuler.net/problem=67
*/

package main

import (
	"bufio"
	"fmt"
	"os"
	"strconv"
	"strings"
)

func max(a, b int) int {
	if a > b {
		return a
	}
	return b
}

func assertNoError(err error) {
	if err != nil {
		fmt.Fprintf(os.Stderr, "Error: %v\n", err)
		os.Exit(1)
	}
}

func run(filename string) {
	var totals []int
	f, err := os.Open(filename)
	assertNoError(err)
	input := bufio.NewScanner(f)

	k := 0
	for input.Scan() {
		var values []int
		for _, s := range strings.Split(input.Text(), " ") {
			value, err := strconv.Atoi(s)
			assertNoError(err)
			values = append(values, value)
		}

		k += 1
		if k == 1 {
			totals = append(totals, values[0])
		} else {
			totals = append(totals, values[k-1]+totals[k-2])
			for i := k - 2; i > 0; i-- {
				totals[i] = values[i] + max(totals[i], totals[i-1])
			}
			totals[0] += values[0]
		}
	}

	maxTotal := 0
	for _, x := range totals {
		if x > maxTotal {
			maxTotal = x
		}
	}
	fmt.Println(maxTotal)
}

func main() {
	run("input18.txt")
	run("input67.txt")
}
